### PR TITLE
boards: arm: add support for STM32F0DISCOVERY

### PR DESCRIPTION
**Pull request by nathantsoi**
Support the ST STM32F0DISCOVERY board with STM32F051R8T6 SoC

Tested with `samples/basic/blinky`

Includes flash config in `stm32f0_disco.rst` for use with mcuboot

Signed-off-by: Nathan Tsoi <nathan@vertile.com>

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Split this into 2 commits, one for the SoC and one for the board.

**Comment by erwango** [comment]
@nathantsoi : Please rebase vs latest master, it will allow to get rid of dts.fixup file.

File: boards/arm/stm32f0_disco/doc/stm32f0_disco.rst [hunk 1]
```diff
@@ -0,0 +1,141 @@
+.. _stm32f0_disco_board:
+
+ST STM32F0DISCOVERY
+################
```
Comment (dbkinder) [inline boards/arm/stm32f0_disco/doc/stm32f0_disco.rst]: heading underline needs to be at least as long as the title (add some more ###)

File: boards/arm/stm32f0_disco/doc/stm32f0_disco.rst [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+Overview
+********
+
+The STM32F0 Discovery development board with STM32F051R8T6 MCU
```
Comment (dbkinder) [inline boards/arm/stm32f0_disco/doc/stm32f0_disco.rst]: Complete sentences are preferred.  Let's combine these into:

    The STM32F0 Discovery development board uses an STM32F051R8T6 MCU and
    integrates the ST-LINK/V2-1 debugger and programmer.  It also comes with a
    comprehensive STM32 software HAL library and various packaged software
    examples.

File: boards/arm/stm32f0_disco/doc/stm32f0_disco.rst [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+Hardware
+********
+
+- STM32F051R8T6 microcontroller featuring 64 KB Flash memory, 8 KB RAM in an LQFP64 package
+- On-board ST-LINK/V2 with selection mode switch to use the kit as a standalone ST-LINK/V2 (with SWD connector for programming and debugging)
+- Board power supply: through USB bus or from an external 5 V supply voltage
+- External application power supply: 3 V and 5 V
+- Four LEDs:
+    - LD1 (red) for 3.3 V power on
```
Comment (dbkinder) [inline boards/arm/stm32f0_disco/doc/stm32f0_disco.rst]: need a blank like before this sub-list

File: boards/arm/stm32f0_disco/doc/stm32f0_disco.rst [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+     :align: center
+     :alt: STM32F0DISCOVERY
+
+More information about the board can be found at the `STM32F0DISCOVERY website`_.
+
+Hardware
+********
+
+- STM32F051R8T6 microcontroller featuring 64 KB Flash memory, 8 KB RAM in an LQFP64 package
+- On-board ST-LINK/V2 with selection mode switch to use the kit as a standalone ST-LINK/V2 (with SWD connector for programming and debugging)
```
Comment (dbkinder) [inline boards/arm/stm32f0_disco/doc/stm32f0_disco.rst]: try to keep line length < 80 characters (unless you've got a long URL):

    - On-board ST-LINK/V2 with selection mode switch to use the kit as a standalone 
      ST-LINK/V2 (with SWD connector for programming and debugging)

File: boards/arm/stm32f0_disco/doc/stm32f0_disco.rst [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+     :width: 350px
+     :height: 404px
+     :align: center
+     :alt: STM32F0DISCOVERY
+
+More information about the board can be found at the `STM32F0DISCOVERY website`_.
+
+Hardware
+********
+
```
Comment (dbkinder) [inline boards/arm/stm32f0_disco/doc/stm32f0_disco.rst]: Let's start with an introduction:

    The STM32 Discovery board features:

    - STM32F051R8T6 microcontroller ...

and make sure there's a blank line before the first list item.

File: boards/arm/stm32f0_disco/doc/stm32f0_disco.rst [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+- Board power supply: through USB bus or from an external 5 V supply voltage
+- External application power supply: 3 V and 5 V
+- Four LEDs:
+    - LD1 (red) for 3.3 V power on
+    - LD2 (red/green) for USB communication
+    - LD3 (green) for PC9 output
+    - LD4 (blue) for PC8 output
+- Two push buttons (user and reset)
+- Extension header for all LQFP64 I/Os for quick connection to prototyping board and easy probing
+- An additional board is provided which can be connected to the extension connector for even easier prototyping and probing.
```
Comment (dbkinder) [inline boards/arm/stm32f0_disco/doc/stm32f0_disco.rst]: break the long lines please

File: boards/arm/stm32f0_disco/doc/stm32f0_disco.rst [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+- Comprehensive free software including a variety of examples, part of STM32CubeF0 package or STSW-STM32049 for legacy Standard Libraries usage
+
+More information about STM32F051R8 can be found here:
```
Comment (dbkinder) [inline boards/arm/stm32f0_disco/doc/stm32f0_disco.rst]: Better to write this as:

    More information about STM32F051R8 can be found in the `STM32F0x8 reference manual`_.

File: boards/arm/stm32f0_disco/doc/stm32f0_disco.rst [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+===================
+
+Each of the GPIO pins can be configured by software as output (push-pull or open-drain), as
+input (with or without pull-up or pull-down), or as peripheral alternate function. Most of the
+GPIO pins are shared with digital or analog alternate functions. All GPIOs are high current
+capable except for analog inputs.
+
+Default Zephyr Peripheral Mapping:
+----------------------------------
+- UART_1_TX : PA9
```
Comment (dbkinder) [inline boards/arm/stm32f0_disco/doc/stm32f0_disco.rst]: add a blank line before the list

**Review by dbkinder — CHANGES_REQUESTED** [review]
Thanks for adding documentation. Here are a few suggested edits.

File: arch/arm/soc/st_stm32/stm32f0/soc.h [hunk 9]
```diff
@@ -55,6 +55,11 @@
 #include <stm32f0xx_ll_spi.h>
 #endif
 
+/* For IMG_MANAGER */
+#if defined(CONFIG_SOC_FLASH_STM32)
```
[the remark this answers is not in the document]
Comment (erwango) [reply arch/arm/soc/st_stm32/stm32f0/soc.h]: #5919 Submitted to get it cleaned across whole tree

File: boards/arm/stm32f0_disco/Kconfig.board [hunk 10]
```diff
@@ -0,0 +1,11 @@
+# Kconfig - STM32F0DISCOVERY development board with STM32F051R8 MCU configuration
+#
+# Copyright (c) 2017 RnDity Sp. z o.o.
```
Comment (erwango) [inline boards/arm/stm32f0_disco/Kconfig.board]: Remove this line

File: boards/arm/stm32f0_disco/Kconfig.defconfig [hunk 11]
```diff
@@ -0,0 +1,34 @@
+# Kconfig - STM32F0 development board with STM32F051R8T6 MCU
+#
+# Copyright (c) 2017 RnDity Sp. z o.o.
```
Comment (erwango) [inline boards/arm/stm32f0_disco/Kconfig.defconfig]: Remove this line

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by nathantsoi** [comment]
i think i have everyone's requested changes in. thanks all for the review and let me know if there is anything else

File: dts/arm/st/stm32f051.dtsi [hunk 12]
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
```
Comment (jmtorrespalma) [inline dts/arm/st/stm32f051.dtsi]: This is inconsistent with other submitted files.
Should probably be:
 `Copyright (c) 2018 Nathan Tsoi`

File: boards/arm/stm32f0_disco/board.h [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+/* USER push button */
+#define USER_PB_GPIO_PORT   "GPIOA"
+#define USER_PB_GPIO_PIN    0
+
+/* LD2 green LED */
+#define GREEN_LED_GPIO_PORT "GPIOC"
+#define GREEN_LED_GPIO_PIN  9
+
+/* LD2 blue LED */
```
Comment (erwango) [inline boards/arm/stm32f0_disco/board.h]: LD4

File: boards/arm/stm32f0_disco/board.h [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+#ifndef __INC_BOARD_H
+#define __INC_BOARD_H
+
+#include <soc.h>
+
+/* USER push button */
+#define USER_PB_GPIO_PORT   "GPIOA"
+#define USER_PB_GPIO_PIN    0
+
+/* LD2 green LED */
```
Comment (erwango) [inline boards/arm/stm32f0_disco/board.h]: LD3

File: boards/arm/stm32f0_disco/Kconfig.defconfig [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+	default y
+
+endif # UART_CONSOLE
+
+# Disable ports D and F to allow kernel test cases common and
+# contex to run on stm32f0_disco.
+# Test cases fail due to 'SRAM' region overflow.
+if GPIO
+
+config GPIO_STM32_PORTD
```
Comment (erwango) [inline boards/arm/stm32f0_disco/Kconfig.defconfig]: Thanks to move GPIO definition in arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.stm32f051x8.
Then disable in stm32f0_disco_defconfig

File: boards/arm/stm32f0_disco/Kconfig.defconfig [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+	default n
+
+config GPIO_STM32_PORTF
```
Comment (erwango) [inline boards/arm/stm32f0_disco/Kconfig.defconfig]: ditto

**Review by erwango — CHANGES_REQUESTED** [review]
Some additional comments otherwise looks good

File: boards/arm/stm32f0_disco/Kconfig.defconfig (hunk 15 above)
Comment (nathantsoi) [reply boards/arm/stm32f0_disco/Kconfig.defconfig]: Thx for the feedback @erwango. I'm not sure I follow though. I think you're asking for something like:

```
diff --git a/boards/arm/stm32f0_disco/Kconfig.defconfig b/boards/arm/stm32f0_disco/Kconfig.defconfig
index 91ccdeda9..f2c417b3a 100644
--- a/boards/arm/stm32f0_disco/Kconfig.defconfig
+++ b/boards/arm/stm32f0_disco/Kconfig.defconfig
@@ -17,17 +17,4 @@ config UART_STM32_PORT_1

 endif # UART_CONSOLE

-# Disable ports D and F to allow kernel test cases common and
-# contex to run on stm32f0_disco.
-# Test cases fail due to 'SRAM' region overflow.
-if GPIO
-
-config GPIO_STM32_PORTD
-       default n
-
-config GPIO_STM32_PORTF
-       default n
-
-endif # GPIO
-
 endif # BOARD_STM32F0_DISCO
diff --git a/boards/arm/stm32f0_disco/stm32f0_disco_defconfig b/boards/arm/stm32f0_disco/stm32f0_disco_defconfig
index b64c9eb14..25dd94a10 100644
--- a/boards/arm/stm32f0_disco/stm32f0_disco_defconfig
+++ b/boards/arm/stm32f0_disco/stm32f0_disco_defconfig
@@ -30,6 +30,8 @@ CONFIG_PINMUX=y

 # GPIO Controller
 CONFIG_GPIO=y
+CONFIG_GPIO_STM32_PORTD=n
+CONFIG_GPIO_STM32_PORTF=n

 # Clock configuration
 CONFIG_CLOCK_CONTROL=y
```

Does that look close?

File: boards/arm/stm32f0_disco/doc/stm32f0_disco.rst (hunk 5 above)
Comment (nathantsoi) [reply boards/arm/stm32f0_disco/doc/stm32f0_disco.rst]: not sure why github didn't pick up the diff, but note that this introduction has been added

**Review by dbkinder — APPROVED** [review]
thanks!

File: boards/arm/stm32f0_disco/Kconfig.defconfig (hunk 15 above)
Comment (erwango) [reply boards/arm/stm32f0_disco/Kconfig.defconfig]: @nathantsoi : Exactly

**Comment by galak** [comment]
@erwango look good to you?

**Review by erwango — APPROVED** [review]
Fine with me!

**Review by galak — APPROVED** [review]

**Pull request merged**